### DISCUSSION TITLE
Are SLAs usable? Are others using them?

### DISCUSSION THREAD

**Discussion by notatallshaw-gts** [apache/airflow#27578] [Q&A]
Since the behavior of timeouts changed for sensors in Airflow 2.3 I've found that we have started adopting SLAs quite a lot more.

The behavior beforehand was to make a sensor that had a timeout of say 1 hour, where we expected it to complete in 1 hour, but allow multiple retries and email on retry. However this flow broke in Airflow 2.3 because when a sensor times out it no longer retries.

After understanding that Airflow SLAs are triggered [relative to the start of a DAG](https://github.com/apache/airflow/pull/27111) we've now started giving sensors long timeouts, a few retries (in case they fail because of external factors), and SLAs of when they expect them to complete by.

We are looking to expend our SLA usage by using the SLA callback to send out alerts to our group chat tool, but I am concerned by the open SLA issues/pull requests on github, e.g.
https://github.com/apache/airflow/pull/27317
https://github.com/apache/airflow/issues/15596

Is there a certain point where we can expect things to break?

**Comment by potiuk** [comment]
SLA is not very useful. Indeed there are various problems with it.

**Reply by notatallshaw-gts** [reply]
I am finding them very useful now that I understand their behavior correctly.

We often have a situation where we want an alert from a sensor if it has not completed by 7am, but we can set the timeout on the sensor to say 12 hours so if it does complete later we don't need to take any manual action.

What I am worried about is that because the community isn't using them enough that that there are many hidden  bugs and problems yet to be identified, and as we start to depend more on SLAs we will run in to these bugs and problems and not be equipped to deal with them.

**Reply by potiuk** [reply]
I think there are various small improvements coming from the member of the community - there a number of newly opened PRs that aim at improving them. https://github.com/apache/airflow/pulls?q=is%3Apr+is%3Aopen+SLA 

I think the best you can do is to engage in those discussions, see what other propose, or even propose improvements and  implement some PRs. 

The "community" is really also you and your company. Airflow has > 2200 contributors and you can become one of them. SLA needs some love - for sure. But if you give it some love, it might be the best way to make it really useful.
